woocommerce/woocommerce-gateway-stripe
# Save payment method during checkout

# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1730 

## Saved payment methods are here at last

The changes in this PR add an implementation to provide functionality to save a payment method during checkout: this attaches the Stripe payment method to the customer's Stripe account and saves a corresponding WC payment token on the store owner's site. Payment methods can then be managed from the customer's account page (deleted or default method changed) and will appear to be used at checkout, so that the customer can check out with ease. 

This PR adds the full flow to save payment methods and reuse them at checkout. In WCPay, when completing a checkout with a saved payment method, we fell back to the parent gateway's functionality. In the Stripe plugin, this was not possible, since the legacy gateway uses Sources instead of Payment Methods and Payment Intents. Consequently, I had to rebuild this flow using Payment Intents instead of Sources: as in WCPay we are still able to create and confirm the intent in one request, but the rest of the flow mirrors the checks and order processing from the existing Stripe gateway.

Available saved payment methods at checkout correspond to payment methods that are enabled in the plugin. Currently only cards are supported, but this PR also actually includes functionality to save payments with Bancontact and iDEAL as well, where the payment method will be converted into a reusable SEPA payment method. You can test this by enabling the `is_reusable` flag on these payment methods and then selecting the save payment method checkbox as you checkout; you should discover that a corresponding SEPA payment method is created that will appear at checkout and in your account, as long as Bancontact and/or iDEAL are also enabled. I have currently disabled them, because since we have not yet implemented SEPA as a payment method, we are unable to complete a checkout using these saved SEPA payment methods--specifically, because [this array of `payment_method_types`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-intent-controller.php#L314) cannot currently contain `sepa_debit` as an option. However, this does currently contain an implementation for the full 3DS authentication flow, so that you should be able to save a [card that requires authentication at checkout](https://stripe.com/docs/testing#regulatory-cards), after proceeding through the [authentication modal](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/classic/upe/index.js#L542), and once this card is saved you will be able to use it as a saved payment method, whilst still being prompted to authenticate the card at checkout via said aforementioned ephemeral modal.

## Unfortunately, some things are not here yet

This PR currently has limited support for subscriptions and Setup Intents, since these have not yet been implemented fully yet with the UPE. I also think there will need to be a little bit of work for this to play nicely with the Pay for Order page, though I have left remnants of the WCPay implementation that may be one day be usable. The existing gateway also contains [some compatability with Pre Orders](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-gateway-stripe.php#L583) and I have left this unimplemented for now within this PR. 

When saving payment methods there are essentially four flows:

1. The card flow.
2. The SEPA flow.
3. The generated SEPA flow (where payment methods support reuse after rebirth as converted SEPA payment methods).
4. The Alipay flow (I have no idea what this one is, since [recurring payments require admin approval from Stripe](https://stripe.com/docs/payments/alipay)).

The flows are generally similar, though they differ most crucially once where we must store these payment methods as WC Payment Tokens, since there are different fields that must be stored and different places inside the Intent response where we must look for these fields. This PR contains an implementation for options 1 and 3: option 2 might work, though I am unable to test without SEPA enabled as a payment method, so it may require a little tinkering once this feature is ready. Alipay is not currently available in the UPE and I'm not sure if there's currently a timeline for it to be added, so I've blissfully ignored that one. 

I also haven't added any tests yet, but I'm going to be AFK for a few days around this upcoming weekend and I think this PR is relatively large, so hopefully someone can tear this PR apart and toss me a cheeky review, before I inevitably get around to adding some test coverage.

# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Pull the changes in this PR and add the appropriate flags to your `wc_options` table to enable the UPE at checkout (I think `_wcstripe_feature_upe`, `woocommerce_stripe_settings.upe_checkout_experience_enabled `, and `_wcstripe_feature_upe_settings` should all have values of `yes`).
2. Add an item from your store to cart and proceed to checkout. Other than any pre-existing functionality, you can now continue to test the following new features. 
3. Save payment method checkbox is only present and selectable for payment methods that are flagged as reusable (`is_reusable` is `true`).
4. Selecting the save payment method checkbox at checkout displays the correct required legal mandate within the UPE at checkout. These mandates remain displayed when they should be, as the checkbox is toggled and selected payment method is changed. 
5. Upon selecting the save payment method checkbox and completing a checkout, the order is completed as expected and the payment method used now appears at checkout as well as in the My Account dashboard. Check that the WC Order has the correct status and notes and confirm in the Stripe dashboard that the payment method used has been successfully set up for future off-session payments. (Optional: test this for payment methods other than cards, by manually setting the `is_reusable` flag for Bancontact and iDEAL.)
6. Customer is able to view all saved payment methods at checkout and is able to complete a checkout by choosing any of them.
7. When completing a checkout with a card that [requires additional 3DS authentication](https://stripe.com/docs/testing#regulatory-cards), the Stripe authentication modal appears at checkout. After the payment method has been saved, if a customer chooses to checkout with this saved payment method, the Stripe authentication modal appears again during checkout, and--once confirmed--proceeds and completes the order as expected. 
8. User can view and delete saved payment methods from the My Account dashboard. (Note that this page has a limit that it will only show 10 saved payment methods or the limit set by `posts_per_page`.)
9. User can change their default saved payment method from the My Account dashboard. This chosen default payment method is also reflected at checkout.
10. I think that's all the test cases I can really think of, but if your imagination is wilder than mine, don't hesitate to entertain it.

-------------------
